### PR TITLE
use docker-compose to spawn a kafka cluster for component tests

### PR DESCRIPTION
**Pull request by DavidSubiros**
### What

Investigate if we can improve component tests for services using kafka: https://trello.com/c/ONk610ze/5151-investigate-implementing-feature-tests-in-dp-import-cantabular-dimension-options-13

There are 3 options:
- Use `dp-kafka/kafkatest` package, but this is just a mock intended for unit tests, which implements the channels logic, wrapping an moq mock.
- Use `sarama.MockBroker` ( - link - )[https://github.com/Shopify/sarama/blob/master/mockbroker.go], but this is just a TCP Server with some framework to provide mock responses that would be returned by the Kafka broker. It also needs the golang testing package (unit tests), which is not exposed by Armakuni framework. The only advantage of using this over dp-kafka/kafkatest is that we would also test dp-kafka library, having the mock at the network level, instead of the channel communication level.
- Spawn a Kafka cluster with docker-compose, and use real producers / consumers with the right configuration. There could be some argument that the last point would be integration test, but I see this approach as a higher level mock (we would not depend on other services, we just use Kafka producers and consumers to trigger and validate tests)

In this PR I implemented the last approach, without any testing validation yet, as a proof of concept. If we agree, we can start using this approach for component tests of services that involve Kafka.

### How to review

- Make sure code changes make sense
- Run `make test-component` and check that there is no panic.
- Check that the CI runs the component tests successfully: `https://concourse.onsdigital.co.uk/teams/main/pipelines/dp-import-cantabular-dimension-options/jobs/pr-component`

### Who can review

Anyone, specially people that were involved with Armakuni component test works

### CODE REVIEW

**Comment by DavidSubiros** [comment]
Note: in order to run docker-compose from the CI docker container, we need to set the `priviledged` flag to true, this is done in the following PRs:
- CI Pipeline confiburable `privileged` flag (defaults to false): https://github.com/ONSdigital/dp-ci/pull/891 
- Set `privileged` flag to true for cantabular import services: https://github.com/ONSdigital/dp-configs/pull/373/files

File: ci/scripts/component.sh [hunk 1]
```diff
@@ -1,5 +1,19 @@
...
+# Run component tests in docker compose defined in features/compose folder
+pushd dp-import-cantabular-dimension-options/features/compose
+  docker-compose up --abort-on-container-exit
+  e=$?
+popd
+
+# Cat the component-test output file and remove it, which is stored int he project folder
 pushd dp-import-cantabular-dimension-options
-  make test-component
+  ls -la
```
Comment (ChrysmOre) [inline ci/scripts/component.sh:11]: Probably don't need this ls anymore

File: features/steps/steps.go [hunk 2]
```diff
@@ -2,87 +2,204 @@ package steps
...
-	content, err := ioutil.ReadFile(c.cfg.OutputFilePath)
+// theInstanceHasDimensionCount generates a mocked response for dataset API
+// GET /instances/{id}/dimensions with the provided totalCount and limit=offset=0
+func (c *Component) theInstanceHasDimensionCount(id string, totalCount string) error {
+
+	// count must be a number
+	cnt, err := strconv.Atoi(totalCount)
 	if err != nil {
-		return err
+		return nil
```
Comment (ChrysmOre) [inline features/steps/steps.go:59]: Why don't you need to return the error here?

File: features/steps/steps.go [hunk 3]
```diff
@@ -2,87 +2,204 @@ package steps
...
+			}
+
+			msg.Commit()
+			msg.Release()
+
+			got = append(got, &e)
+		}
+	}
+
+	if !reflect.DeepEqual(got, expected) {
```
Comment (ChrysmOre) [inline features/steps/steps.go:162]: I also realised that printing here as it was wasn't good enough as you only got the addresses, however using `github.com/google/go-cmp/cmp` here may be a better option (which is what I went for), the comparison ends up a one liner and does everything for us but all type safe etc
```
	if diff := cmp.Diff(got, expected); diff != "" {
		return fmt.Errorf("-got +expected)\n%s\n", diff)
	}
```

**Review by ChrysmOre — COMMENTED** [review]
Couple of minor comments, mostly looks good though

File: features/steps/steps.go (hunk 2 above)
Comment (DavidSubiros) [reply features/steps/steps.go:59]: I don't know, I think this should be an error, I will change it to return the error.

File: features/steps/steps.go (hunk 3 above)
Comment (DavidSubiros) [reply features/steps/steps.go:162]: ok, I will change it. I added this method to do it:
```
func printValues(in []*event.InstanceComplete) string {
	ret := "[ "
	for _, val := range in {
		ret += fmt.Sprintf("%+v ", val)
	}
	return ret + "]"
}
```
But maybe it's better to use go-cmp

File: features/steps/steps.go [hunk 4]
```diff
@@ -2,87 +2,194 @@ package steps
...
+}
+
+// theInstanceHasDimensionCount generates a mocked response for dataset API
+// GET /instances/{id}/dimensions with the provided totalCount and limit=offset=0
+func (c *Component) theInstanceHasDimensionCount(id string, totalCount string) error {
+
+	// count must be a number
+	cnt, err := strconv.Atoi(totalCount)
 	if err != nil {
 		return err
```
Comment (ChrysmOre) [inline features/steps/steps.go:59]: Probably want some Errorf context here?

**Review by ChrysmOre — CHANGES_REQUESTED** [review]
One tiny comment otherwise LGTM

**Review by ChrysmOre — APPROVED** [review]

**Pull request merged**